Introduce AutoGeneratedTranscriptionManager and AutoGeneratedTranscriptionManagerProxy

danae404 opened the pull request:
#### 58234436293ff4f4843cf8d62084289d2319dd6d
<pre>
Introduce AutoGeneratedTranscriptionManager and AutoGeneratedTranscriptionManagerProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=281701">https://bugs.webkit.org/show_bug.cgi?id=281701</a>
<a href="https://rdar.apple.com/138150093">rdar://138150093</a>

Reviewed by NOBODY (OOPS!).

Based on work by Osose Ewaleifoh.

This change introduces two new classes: AutoGeneratedTranscriptionManagerProxy in the web process
(owned by WebProcess) and AutoGeneratedTranscriptionManager in the UI process (owned by WebProcessProxy).
Currently, these classes only contain stubs. They will be used in future changes to perform video
transcription.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::playPlayer):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::startTranscription):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::startTranscription):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Media/AutoGeneratedTranscriptionManager.cpp: Added.
(WebKit::AutoGeneratedTranscriptionManager::AutoGeneratedTranscriptionManager):
(WebKit::AutoGeneratedTranscriptionManager::registerClient):
(WebKit::AutoGeneratedTranscriptionManager::unregisterClient):
(WebKit::AutoGeneratedTranscriptionManager::audioStorageChanged):
(WebKit::AutoGeneratedTranscriptionManager::audioSamplesAvailable):
(WebKit::AutoGeneratedTranscriptionManager::startTranscription):
(WebKit::AutoGeneratedTranscriptionManager::stopTranscription):
(WebKit::AutoGeneratedTranscriptionManager::createAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManager::updateAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManager::finalizeAutoGeneratedCaptionCue):
* Source/WebKit/UIProcess/Media/AutoGeneratedTranscriptionManager.h: Added.
(WebKit::AutoGeneratedTranscriptionManager::~AutoGeneratedTranscriptionManager):
* Source/WebKit/UIProcess/Media/AutoGeneratedTranscriptionManager.messages.in: Added.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::autoGeneratedTranscriptionManager):
(WebKit::WebProcessProxy::ensureAutoGeneratedTranscriptionManager):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::startTranscription):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::autoGeneratedTranscriptionManagerProxy):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/AutoGeneratedTranscriptionManagerProxy.cpp: Added.
(WebKit::AutoGeneratedTranscriptionManagerProxy::AutoGeneratedTranscriptionManagerProxy):
(WebKit::AutoGeneratedTranscriptionManagerProxy::startTranscription):
(WebKit::AutoGeneratedTranscriptionManagerProxy::stopTranscription):
(WebKit::AutoGeneratedTranscriptionManagerProxy::createAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManagerProxy::updateAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManagerProxy::finalizeAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManagerProxy::registerClient):
(WebKit::AutoGeneratedTranscriptionManagerProxy::unregisterClient):
* Source/WebKit/WebProcess/cocoa/AutoGeneratedTranscriptionManagerProxy.h: Added.
(WebKit::AutoGeneratedTranscriptionManagerProxy::~AutoGeneratedTranscriptionManagerProxy):
(WebKit::AutoGeneratedTranscriptionManagerProxy::connection):
* Source/WebKit/WebProcess/cocoa/AutoGeneratedTranscriptionManagerProxy.messages.in: Added.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58234436293ff4f4843cf8d62084289d2319dd6d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73337 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52766 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26145 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77572 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24575 "Hash 58234436 for PR 35391 does not build (failure)") 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60572 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/549 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/77572 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/24575 "Hash 58234436 for PR 35391 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76404 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/60572 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/26145 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/77572 "Hash 58234436 for PR 35391 does not build (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/60572 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/26145 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22904 "Hash 58234436 for PR 35391 does not build (failure)") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/60572 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/26145 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79219 "Hash 58234436 for PR 35391 does not build (failure)") | 
| | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/652 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/549 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/79219 "Hash 58234436 for PR 35391 does not build (failure)") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/794 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/26145 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/79219 "Hash 58234436 for PR 35391 does not build (failure)") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/55/builds/26145 "Hash 58234436 for PR 35391 does not build (failure)") | | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/616 "Hash 58234436 for PR 35391 does not build (failure)") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3353 "Failed to build and analyze WebKit") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/646 "Hash 58234436 for PR 35391 does not build (failure)") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/630 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/648 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6469,6 +6469,11 @@ void HTMLMediaElement::playPlayer()
+#if HAVE(SPEECHRECOGNIZER)
+    if (document().settings().autoGeneratedTranscriptionForVideoEnabled())
+        player->startTranscription();
+#endif
```
We don't want to start transcription _every_ time a player starts, instead it should only start when the transcription track created by a player is enabled. The logic to do this will be spread across several classes, so why don't you just remove all of the `startTranscription` code in `HTMLMediaElement` and the `MediaPlayer*` classes for now.

danae404 replied:
I updated the patch with the startTranscription() code removed, but i wonder if it could make sense to call startTranscription when the video is loaded or played for the first time, and within the startTranscription code we could create the synthesized text track and have it disabled by default

danae404 replied:
We talked about this over slack and decided that having startTranscription() be called when the video is loaded would have an impact on power and performance - it would be best to start and stop transcription as the text track is enabled and disabled

eric-carlson commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1314,6 +1316,20 @@ void WebProcessProxy::processDidTerminateOrFailedToLaunch(ProcessTerminationReas
+        m_autoGeneratedTranscriptionManager = WTF::makeUnique<AutoGeneratedTranscriptionManager>(&protectedConnection().get());
+        addMessageReceiver(Messages::AutoGeneratedTranscriptionManager::messageReceiverName(), *m_autoGeneratedTranscriptionManager);
+    }
+}
```
Unless `ensureAutoGeneratedTranscriptionManager` will be used in other methods, you might as well combine these.

eric-carlson commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -191,6 +192,10 @@ class WebProcessProxy final : public AuxiliaryProcessProxy {
+
+    void ensureAutoGeneratedTranscriptionManager();
+    AutoGeneratedTranscriptionManager&
+    autoGeneratedTranscriptionManager();
```
Unintentional line break?

danae404 committed:
Introduce AutoGeneratedTranscriptionManager and AutoGeneratedTranscriptionManagerProxy
https://bugs.webkit.org/show_bug.cgi?id=281701
rdar://138150093

Reviewed by NOBODY (OOPS!).

Based on work by Osose Ewaleifoh.

This change introduces two new classes: AutoGeneratedTranscriptionManagerProxy in the web process
(owned by WebProcess) and AutoGeneratedTranscriptionManager in the UI process (owned by WebProcessProxy).
Currently, these classes only contain stubs. They will be used in future changes to perform video
transcription.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::playPlayer):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::startTranscription):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::startTranscription):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Media/AutoGeneratedTranscriptionManager.cpp: Added.
(WebKit::AutoGeneratedTranscriptionManager::AutoGeneratedTranscriptionManager):
(WebKit::AutoGeneratedTranscriptionManager::registerClient):
(WebKit::AutoGeneratedTranscriptionManager::unregisterClient):
(WebKit::AutoGeneratedTranscriptionManager::audioStorageChanged):
(WebKit::AutoGeneratedTranscriptionManager::audioSamplesAvailable):
(WebKit::AutoGeneratedTranscriptionManager::startTranscription):
(WebKit::AutoGeneratedTranscriptionManager::stopTranscription):
(WebKit::AutoGeneratedTranscriptionManager::createAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManager::updateAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManager::finalizeAutoGeneratedCaptionCue):
* Source/WebKit/UIProcess/Media/AutoGeneratedTranscriptionManager.h: Added.
(WebKit::AutoGeneratedTranscriptionManager::~AutoGeneratedTranscriptionManager):
* Source/WebKit/UIProcess/Media/AutoGeneratedTranscriptionManager.messages.in: Added.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::autoGeneratedTranscriptionManager):
(WebKit::WebProcessProxy::ensureAutoGeneratedTranscriptionManager):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::startTranscription):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::autoGeneratedTranscriptionManagerProxy):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/AutoGeneratedTranscriptionManagerProxy.cpp: Added.
(WebKit::AutoGeneratedTranscriptionManagerProxy::AutoGeneratedTranscriptionManagerProxy):
(WebKit::AutoGeneratedTranscriptionManagerProxy::startTranscription):
(WebKit::AutoGeneratedTranscriptionManagerProxy::stopTranscription):
(WebKit::AutoGeneratedTranscriptionManagerProxy::createAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManagerProxy::updateAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManagerProxy::finalizeAutoGeneratedCaptionCue):
(WebKit::AutoGeneratedTranscriptionManagerProxy::registerClient):
(WebKit::AutoGeneratedTranscriptionManagerProxy::unregisterClient):
* Source/WebKit/WebProcess/cocoa/AutoGeneratedTranscriptionManagerProxy.h: Added.
(WebKit::AutoGeneratedTranscriptionManagerProxy::~AutoGeneratedTranscriptionManagerProxy):
(WebKit::AutoGeneratedTranscriptionManagerProxy::connection):
* Source/WebKit/WebProcess/cocoa/AutoGeneratedTranscriptionManagerProxy.messages.in: Added.
